### PR TITLE
Add ecs_service tasks placement constraints and strategy.

### PR DESCRIPTION
**Pull request by anryko**
##### ISSUE TYPE

 - Feature Pull Request

##### COMPONENT NAME

module_name: ecs_service

##### ANSIBLE VERSION

```
ansible 2.3.0 (devel 75a7ebe900) last updated 2017/01/19 17:01:06 (GMT +200)
```

##### SUMMARY

Make use of AWS ECS task placement constraints and strategy functionality. You can restrict your task to run only on specific ec2 instances of ECS cluster, according to specified placement strategy.

Before change:
```
fatal: [localhost]: FAILED! => {
    "changed": false,
    "failed": true,
    "invocation": {
        "module_args": {
            "cluster": "dev-ecs-cluster-1",
            "desired_count": "1",
            "load_balancers": [
                {
                    "containerName": "web-app-test",
                    "containerPort": "8080",
                    "loadBalancerName": "dev-web-app-test-lb"
                }
            ],
            "name": "web-app-test-service",
            "placement_constraints": [
                {
                    "expression": "attribute:flavor==cpu",
                    "type": "memberOf"
                }
            ],
            "placement_strategy": [
                {
                    "field": "memory",
                    "type": "binpack"
                }
            ],
            "role": "ecsServiceRole",
            "state": "present",
            "task_definition": "dev-web-app-test-task-2f4255c"
        },
        "module_name": "ecs_service"
    },
    "msg": "Unsupported parameters for (ecs_service) module: placement_constraints. Supported parameters include: aws_access_key,aws_secret_key,client_token,cluster,delay,desired_count,ec2_url,load_balancers,name,profile,region,rep
eat,role,security_token,state,task_definition,validate_certs"
}
```

After change:
```
changed: [localhost] => {
    "changed": true,
    "invocation": {
        "module_args": {
            "aws_access_key": null,
            "aws_secret_key": null,
            "client_token": "",
            "cluster": "dev-ecs-cluster-1",
            "delay": 10,
            "desired_count": 1,
            "ec2_url": null,
            "load_balancers": [
                {
                    "containerName": "web-app-test",
                    "containerPort": 8080,
                    "loadBalancerName": "dev-web-app-test-lb"
                }
            ],
            "name": "web-app-test-service",
            "placement_constraints": [
                {
                    "expression": "attribute:flavor==cpu",
                    "type": "memberOf"
                }
            ],
            "placement_strategy": [
                {
                    "field": "memory",
                    "type": "binpack"
                }
            ],
            "profile": null,
            "region": null,
            "repeat": 10,
            "role": "ecsServiceRole",
            "security_token": null,
            "state": "present",
            "task_definition": "dev-web-app-test-task-2f4255c",
            "validate_certs": true
        },
        "module_name": "ecs_service"
    },
    "service": {
        "clusterArn": "arn:aws:ecs:eu-central-1:000000000000:cluster/dev-ecs-cluster-1",
        "createdAt": "2017-01-20T15:11:00.880000+01:00",
        "deploymentConfiguration": {
            "maximumPercent": 200,
            "minimumHealthyPercent": 100
        },
        "deployments": [
            {
                "createdAt": "2017-01-20 15:11:00.880000+01:00",
                "desiredCount": 1,
                "id": "ecs-svc/0000000000000000000",
                "pendingCount": 0,
                "runningCount": 1,
                "status": "PRIMARY",
                "taskDefinition": "arn:aws:ecs:eu-central-1:000000000000:task-definition/dev-web-app-test-task-2f4255c:1",
                "updatedAt": "2017-01-20 15:11:45.071000+01:00"
            }
        ],
        "desiredCount": 1,
        "events": [
        ],
        "loadBalancers": [
            {
                "containerName": "web-app-test",
                "containerPort": 8080,
                "loadBalancerName": "dev-web-app-test-lb"
            }
        ],
        "pendingCount": 0,
        "placementConstraints": [
            {
                "expression": "attribute:flavor==cpu",
                "type": "memberOf"
            }
        ],
        "placementStrategy": [
            {
                "field": "MEMORY",
                "type": "binpack"
            }
        ],
        "roleArn": "arn:aws:iam::000000000000:role/ecsServiceRole",
        "runningCount": 1,
        "serviceArn": "arn:aws:ecs:eu-central-1:000000000000:service/web-app-test-service",
        "serviceName": "web-app-test-service",
        "status": "ACTIVE",
        "taskDefinition": "arn:aws:ecs:eu-central-1:000000000000:task-definition/dev-web-app-test-task-2f4255c:1"
    }
}
```

### CODE REVIEW

**Comment by simplesteph** [comment]
Shipit

**Comment by simplesteph** [comment]
Good stuff this will be definitely needed. Also thanks for the fix on type casting for the load balancer port

**Comment by simplesteph** [comment]
Hi can you please include the placement constraints in the update function as well?

**Comment by anryko** [comment]
@simplesteph unfortunately boto3 [update_service](http://boto3.readthedocs.io/en/latest/reference/services/ecs.html#ECS.Client.update_service) doesn't support placementConstraints parameter.

**Comment by simplesteph** [comment]
That's kind of crap. Can we circumvent the problem by deleting then
re-creating the service in case of a change in that parameter ?

Thoughts ?

On 21 Jan. 2017 9:31 am, "Andrej Svenke" <notifications@github.com> wrote:

@simplesteph <https://github.com/simplesteph> unfortunately boto3
update_service
<http://boto3.readthedocs.io/en/latest/reference/services/ecs.html#ECS.Client.update_service>
doesn't support placementConstraints parameter.

—
You are receiving this because you were mentioned.
Reply to this email directly, view it on GitHub
<https://github.com/ansible/ansible/pull/20503#issuecomment-274197507>, or mute
the thread

**Comment by simplesteph** [comment]
Scratch that deleting the service would delete tasks. I'm a bit tired haha
:)

On 22 Jan. 2017 2:55 pm, wrote:

That's kind of crap. Can we circumvent the problem by deleting then
re-creating the service in case of a change in that parameter ?

Thoughts ?

On 21 Jan. 2017 9:31 am, "Andrej Svenke" <notifications@github.com> wrote:

@simplesteph <https://github.com/simplesteph> unfortunately boto3
update_service
<http://boto3.readthedocs.io/en/latest/reference/services/ecs.html#ECS.Client.update_service>
doesn't support placementConstraints parameter.

—
You are receiving this because you were mentioned.
Reply to this email directly, view it on GitHub
<https://github.com/ansible/ansible/pull/20503#issuecomment-274197507>, or mute
the thread

**Comment by simplesteph** [comment]
Alright it's not supported through the AWS console either so really no hope here. 
Can we support `placementStrategy` parameter as part of that PR too please? the two kinda go along together.

**Review by simplesteph — APPROVED** [review]

**Comment by simplesteph** [comment]
Shipit

**Comment by schwartzmx** [comment]
bot_broken

not sure why I'm getting @'d on this

**Comment by anryko** [comment]
Bot is fine. I broke my fork and it changed the PR with all the stuff i merged from upstream. Then I quickly fixed it, however bot was faster...

**Comment by schwartzmx** [comment]
I see :),  thanks for the clarification @anryko 👍

**Comment by simplesteph** [comment]
Shipit

**Comment by simplesteph** [comment]
@anryko quick conflicts to resolve due to my merge, and we're pushing to ship this! Thanks heaps

**Comment by simplesteph** [comment]
@anryko quick reminder to rebase to resolve conflicts. Then I'll shipit

**Comment by anryko** [comment]
@simplesteph ready

**Comment by simplesteph** [comment]
Shipit

**Comment by simplesteph** [comment]
@bcoca unsure why the bot isn't picking my shipit...

shipit

**Comment by simplesteph** [comment]
@rmorlok It seems we now need 2 shipit for the bot to pickup the shipit flags.... Can you please review and approve if possible?

**Comment by simplesteph** [comment]
bot_status

**Comment by simplesteph** [comment]
@ryanbsaunders @wimnat @bcoca @willthames I'm the only active maintainer on these ecs modules. I think there's a minimum of two shipits before stuff gets tagged correctly? How can we proceed moving forward? I don't mind pinging you all the time but you surely would mind

**Comment by willthames** [comment]
@simplesteph feel free to keep pinging me. I'm the fallback maintainer on all aws modules after all.

**Comment by willthames** [comment]
shipit - this change is suitably simple and elegant.

**Comment by simplesteph** [comment]
@willthames anything more to do to merge this?

**Comment by willthames** [comment]
hmm, nothing should be needed, I'd think it should just get merged automatically now - I'll ask on #ansible-devel to see why that hasn't happened

File: lib/ansible/modules/cloud/amazon/ecs_service.py
Comment (bcoca) [inline lib/ansible/modules/cloud/amazon/ecs_service.py]: missing version_added

File: lib/ansible/modules/cloud/amazon/ecs_service.py
Comment (bcoca) [inline lib/ansible/modules/cloud/amazon/ecs_service.py]: should note any example using this only works on 2.3 and above

**Review by bcoca — CHANGES_REQUESTED** [review]

**Comment by anryko** [comment]
@bcoca thanks! fixed.

**Comment by ryanbsaunders** [comment]
@simplesteph perhaps you are looking for someone else.  I'm not a contributor to this project.

**Comment by simplesteph** [comment]
@willthames shipit

**Comment by simplesteph** [comment]
ping @willthames

**Comment by simplesteph** [comment]
ping @gundalow @willthames . This one has been outstanding for too long

File: lib/ansible/modules/cloud/amazon/ecs_service.py
Comment (gundalow) [inline lib/ansible/modules/cloud/amazon/ecs_service.py]: `version_added: "2.4"`

File: lib/ansible/modules/cloud/amazon/ecs_service.py
Comment (gundalow) [inline lib/ansible/modules/cloud/amazon/ecs_service.py]: `version_added: "2.4"`

File: lib/ansible/modules/cloud/amazon/ecs_service.py
Comment (gundalow) [inline lib/ansible/modules/cloud/amazon/ecs_service.py]: 2.4

**Review by gundalow — CHANGES_REQUESTED** [review]

**Comment by gundalow** [comment]
Feature freeze for 2.3 was many moons ago, so this will go into 2.4

So just one set of updates then this can be merged into `devel`

**Comment by anryko** [comment]
@gundalow done

**Comment by gundalow** [comment]
Thanks for the prompt fix, almost there:
```
lib/ansible/modules/cloud/amazon/ecs_service.py:232:161: E501 line too long (178 > 160 characters)
lib/ansible/modules/cloud/amazon/ecs_service.py:245:161: E501 line too long (371 > 160 characters)
```

**Comment by anryko** [comment]
@gundalow ready

**Comment by simplesteph** [comment]
Ping @gundalow @winmat

**Review by gundalow — APPROVED** [review]

**Comment by simplesteph** [comment]
shipit

**Comment by simplesteph** [comment]
Ping @gundalow @winmat @bcoca 
Can you please merge this? It has been opened for so long now...

**Review by gundalow — APPROVED** [review]

**Comment by simplesteph** [comment]
Ping @gundalow @winmat @bcoca
it's getting tiring... I'm here to maintain a few modules, do my best to promptly reply and review any PR, and any issue, and this one has been outstanding for way too long. It's not encouraging for the community to submit a simple PR and wait for ages to get it merged. I really don't know what's up. The two other maintainers for the ecs_ modules are completely out to my knowledge, and that means I'll never get two shipit to get the bot to kick in. 
Seriously, to me this issue is a symptom of a much bigger issue... Ansible is an awesome project, but I think there's a way too long lead time to get anything merged in. I'm waiting on this PR to get merged before I do my own, and I'm already afraid it will take just as long to get it through

**Review by gundalow — APPROVED** [review]

**Comment by gundalow** [comment]
@simplesteph thank you for your comments, I understand your comments, and the concerns around the PR backlog are shared by the core team. We are doing various things internally to improve this. 

Maintainers are defined in https://github.com/ansible/ansibullbot/blob/master/MAINTAINERS.txt

If you are willing, it would be great to get yourself added to some of the modules, just raise a PR to update the file.

**Comment by simplesteph** [comment]
@gundalow Thanks for merging the PR. I am already added to the modules that I'm interested of maintaining (`ecs_`), see 
https://github.com/ansible/ansibullbot/blob/master/MAINTAINERS.txt#L52

What I'm wondering is what the maintainer power grants me? Because it seems I can't flip the bot to shipit mode, I can't merge anything, etc. Am I missing something?

**Comment by gundalow** [comment]
@simplesteph Have a look at https://github.com/ansible/ansibullbot/blob/master/ISSUE_HELP.md

Basically if you are a maintainer for a module (or directory) it means your `shipits` count, as detailed in the above doc, shipits from two maintainers mean the module will get a `label:shipit`

**Comment by simplesteph** [comment]
@gundalow that's my understanding as well, but the two other maintainers @rmorlok and @java1guy aren't active to my knowledge, therefore I can't achieve anything

**Comment by gundalow** [comment]
Ah, see what you mean I've raised https://github.com/ansible/ansibullbot/issues/506 to track this. 

Of a module has a single maintainer we will message maintainers of modules on the same sub-topic

Topic = cloud
Sub-topic = amazon

**Pull request merged**
